Lecture 4: cart

HidJaba-hub opened the pull request:
Lecture 4: cart

HidJaba-hub committed:
-fixed .jsp

Lecture 4: cart

HidJaba-hub committed:
Lecture 4: cart

HidJaba-hub committed:
-fixed .jsp

HidJaba-hub committed:
Merge remote-tracking branch 'origin/lecture-4' into lecture-4

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +51,59 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+        synchronized (syncObject) {
+            Product product = productService.getProductById(productId);
+            Optional<CartItem> optionalCartItem = findCartItem(cart, product);
+            checkQuantity(product, quantity, 0);
```
why 0

HidJaba-hub replied:
Because in update method we don't need amount of products in cart (CartQuantity). So to escape code duplication i decided to pass zero CartQuantity

esSoftPalinaTsi replied:
then overriding may be used
0 looks weird tbh

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +51,59 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+            if (optionalCartItem.isPresent()) {
+                if (optionalCartItem.get().getQuantity() == quantity) {
+                    throw new SameArgumentException(optionalCartItem.get());
+                    //it is for resource economy: if the quantity is the same there is no sense to waste our resources
```
why customer can't add 1 iphone and then some time later add one more? we are in e-commerce, where the main aim is to give the ability to sell the goods

HidJaba-hub replied:
it is in update method, in "add" we still can add product with the same quntity. When we click button "update" all the products are going to be updated. With the same quantity too: like we had 1 product and now it is still 1, and it will be updated and i suppose it is not so optimized. For example if it will be database there is no need to access it. So i decided to create this custom exception to show if the products are really updated or not, and if none of products are updated there won't be message "successfully updated"

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +51,59 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+        boolean isDeleted;
+        Object syncObject = SyncObjectPool.getSyncObject(cart.getId().toString());
+        synchronized (syncObject) {
+            isDeleted = cart.getItems().removeIf(cartItem ->
```
why do we need to store some result marker?
we can simply check if we can delete -> if yes then delete, if no throw your exception

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +51,59 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+                .sum());
+
+        cart.setTotalPrice(BigDecimal.valueOf(cart.getItems().stream()
+                .mapToInt(item -> item.getProduct().getPrice().intValueExact() * item.getQuantity())
```
cart.totalPrice should be BigDecimal
int is smaller than BigDecimal, no need to map to int

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/AddCartItemServlet.java:
```diff
@@ -0,0 +1,42 @@
+        String[] productIds = request.getParameterValues("productId");
+        String[] quantities = request.getParameterValues("quantity");
+        String redirectionPath = request.getParameter("path");
+        Long productId = Long.valueOf(request.getPathInfo().substring(1));
```
NPE

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/AddCartItemServlet.java:
```diff
@@ -0,0 +1,42 @@
+
+        int quantity;
+        try {
+            quantity = parseQuantity(quantities[index], request);
```
please create custom validator before parsing the quantity

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/CartPageServlet.java:
```diff
@@ -0,0 +1,50 @@
+
+            int quantity;
+            try {
+                quantity = parseQuantity(quantities[i], request);
```
validate before parsing

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/DeleteCartItemServlet.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        long productId = Long.valueOf(request.getPathInfo().substring(1));
```
NPE

HidJaba-hub committed:
Fixed issues:
-added minicart
-fixed bigdecimal in recalculate method

HidJaba-hub committed:
Fixed issues:
-added minicart
-fixed bigdecimal in recalculate method

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +51,55 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+
+            if (optionalCartItem.isPresent()) {
+                if (optionalCartItem.get().getQuantity() == quantity) {
+                    throw new SameArgumentException(optionalCartItem.get());
```
we can just ignore those products which should not be updated, exceptions are necessary for the exceptional cases
when something goes wrong
in this case I don't see any exceptional case

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/utils/ParseValidator.java:
```diff
@@ -0,0 +1,12 @@
+package com.es.phoneshop.utils;
+
+public class ParseValidator {
```
generally it should be a singleton and implement some interface
errorMap should be a parameter of such Validator and it can be updated here in case of invalid input

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/MiniCartServlet.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+        request.setAttribute("cart", cartService.getCart(request));
+        request.getRequestDispatcher("/WEB-INF/pages/minicart.jsp").include(request, response);
```
why mini cart is not displaying when some error case was triggered

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/pages/cart.jsp:
```diff
@@ -0,0 +1,97 @@
+<c:set var="servletContextPath" value="${pageContext.servletContext.contextPath}"/>
+<tags:master pageTitle="Cart">
+    <p>
+        Cart: ${cart}
```
now no need to display this object in this way, can be removed from al jsp-s
miniCart and cart page is enough

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/pages/minicart.jsp:
```diff
@@ -0,0 +1,10 @@
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+<%@ taglib prefix="tags" tagdir="/WEB-INF/tags" %>
+
+<jsp:useBean id="cart" type="com.es.phoneshop.model.entity.cart.Cart" scope="request"/>
```
request?

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/web.xml:
```diff
@@ -50,6 +51,46 @@
+
+    <servlet-mapping>
+        <servlet-name>miniCart</servlet-name>
+        <url-pattern>/cart/minicart</url-pattern>
```
why /cart/minicart

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/exception/SameArgumentException.java:
```diff
@@ -0,0 +1,12 @@
+
+import com.es.phoneshop.model.entity.cart.CartItem;
+
+public class SameArgumentException extends RuntimeException {
```
if the argument is the same, then it's not an exceptional case

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/utils/ReferenceTool.java:
```diff
@@ -0,0 +1,18 @@
+package com.es.phoneshop.utils;
+
+public class ReferenceTool<T> {
```
what for?

HidJaba-hub replied:
this is for the counter of non-updated products. Because i needed to use reference-typed object (returning of the int counter is not a good variant) that will save counter from method to method. I used List<Integer>, but referenced integer i suppose is better

HidJaba-hub committed:
Fixed issues:
-minicart
-validator

HidJaba-hub committed:
Merge remote-tracking branch 'origin/lecture-4' into lecture-4

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/utils/ParseValidator.java:
```diff
@@ -3,10 +3,14 @@
             return false;
         }
     }
+    public static String removeMinusesFromString(String input) {
```
validator is not changing the content which it is validating
it says if the input is valid or not

HidJaba-hub committed:
Fixed issues:
-reformated
-improved ParseValidator
-renamed methods
-added singleton validator
-fixed missed cart

HidJaba-hub committed:
Fixed issues:
-added singleton validator
-fixed missed cart

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +50,56 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+                if (optionalCartItem.get().getQuantity() != quantity) {
+                    optionalCartItem.get().setQuantity(quantity);
+                    recalculateCart(cart);
+                    return true;
```
spagetti code
no need to store the result of update/save/delete/etc

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +50,56 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+    }
+
+    @Override
+    public void deleteProductInCart(Cart cart, Long productId) throws ProductNotFoundException {
```
deleteProduct

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/CartService.java:
```diff
@@ -9,4 +9,8 @@ public interface CartService {
     Cart getCart(HttpServletRequest request);
 
     void addProductToCart(Cart cart, Long productId, int quantity) throws OutOfStockException;
```
addProduct

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/CartService.java:
```diff
@@ -9,4 +9,8 @@ public interface CartService {
 
     void addProductToCart(Cart cart, Long productId, int quantity) throws OutOfStockException;
+
+    boolean updateProductInCart(Cart cart, Long productId, int quantity) throws OutOfStockException;
```
updateProduct

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/CartService.java:
```diff
@@ -9,4 +9,8 @@ public interface CartService {
+
+    boolean updateProductInCart(Cart cart, Long productId, int quantity) throws OutOfStockException;
+
+    void deleteProductInCart(Cart cart, Long productId);
```
deleteProduct

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/CartItemServlet.java:
```diff
@@ -0,0 +1,63 @@
+        Cart cart = cartService.getCart(request);
+        try {
+            if (!cartService.updateProductInCart(cart, productId, quantity)) {
+                sameQuantityCount.set(sameQuantityCount.get() + 1);
```
Servlet is done to be able to show data, data counting/modification is up to Service layer

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/utils/ParseValidator.java:
```diff
@@ -0,0 +1,8 @@
+package com.es.phoneshop.utils;
```
Validator is not util

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/CartPageServlet.java:
```diff
@@ -0,0 +1,54 @@
+
+            int quantity;
+            try {
+                if (quantityParseValidator.validate(quantities[i], errors, productId)) {
```
quantityParseValidator.validate();
if (!errorMap.isEmpty()) -> parseQuantity

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -52,48 +47,32 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
             int quantity;
             try {
-                quantity = parseQuantity(request);
+                quantity = parseQuantity(quantityStr, request);
```
validator and errorMap should be applied here too

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/cart/CartPageServlet.java:
```diff
@@ -0,0 +1,54 @@
+                errors.put(productId, "Not a number");
+                continue;
+            }
+            updateProductToCart(request, productId, quantity, errors, sameQuantityCount);
```
updateProduct

HidJaba-hub committed:
Fixed issues:
-improved ParseValidator
-renamed methods

HidJaba-hub committed:
Fixed issues:
-reformated

HidJaba-hub committed:
Merge remote-tracking branch 'origin/lecture-4' into lecture-4

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -48,9 +49,54 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
+    private void recalculateCart(Cart cart) {
+        cart.setTotalQuantity(cart.getItems().stream()
+                .map(CartItem::getQuantity)
+                .mapToInt(cartItem -> cartItem)
```
.mapToInt(Integer::intValue) will look better

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/web.xml:
```diff
@@ -50,6 +51,46 @@
+
+    <servlet-mapping>
+        <servlet-name>miniCart</servlet-name>
+        <url-pattern>/minicart/*</url-pattern>
```
do we really need /* ?

HidJaba-hub committed:
Fixed issues:
-IntValue
-removed *
